modules/kafka/src/main/java/org/testcontainers/containers/KafkaContainer.java

```diff
@@ -57,67 +70,161 @@ public KafkaContainer(final DockerImageName dockerImageName) {
         withEnv("KAFKA_TRANSACTION_STATE_LOG_MIN_ISR", DEFAULT_INTERNAL_TOPIC_RF);
         withEnv("KAFKA_LOG_FLUSH_INTERVAL_MESSAGES", Long.MAX_VALUE + "");
         withEnv("KAFKA_GROUP_INITIAL_REBALANCE_DELAY_MS", "0");
+
+        withExposedPorts(KAFKA_PORT);
+
+        withCreateContainerCmdModifier(cmd -> {
+            cmd.withEntrypoint("sh");
+        });
+        withCommand("-c", "while [ ! -f " + STARTER_SCRIPT + " ]; do sleep 0.1; done; " + STARTER_SCRIPT);
     }
 
     public KafkaContainer withEmbeddedZookeeper() {
+        if (this.kraftEnabled) {
+            throw new IllegalStateException("Cannot configure Zookeeper when using Kraft mode");
+        }
         externalZookeeperConnect = null;
         return self();
     }
 
     public KafkaContainer withExternalZookeeper(String connectString) {
+        if (this.kraftEnabled) {
+            throw new IllegalStateException("Cannot configure Zookeeper when using Kraft mode");
+        }
         externalZookeeperConnect = connectString;
         return self();
     }
 
+    public KafkaContainer withKraft() {
+        if (this.externalZookeeperConnect != null) {
+            throw new IllegalStateException("Cannot configure Kraft mode when Zookeeper configured");
+        }
+        verifyMinKraftVersion();
+        kraftEnabled = true;
+        return self();
+    }
+
+    private void verifyMinKraftVersion() {
+        String actualVersion = DockerImageName.parse(getDockerImageName()).getVersionPart();
+        if (new ComparableVersion(actualVersion).isLessThan(MIN_KRAFT_TAG)) {
+            throw new IllegalArgumentException(
+                String.format(
+                    "Provided Confluent Platform's version %s is not supported in Kraft mode (must be %s or above)",
+                    actualVersion,
+                    MIN_KRAFT_TAG
+                )
+            );
+        }
+    }
+
+    public KafkaContainer withClusterId(String clusterId) {
+        this.clusterId = clusterId;
+        return self();
+    }
+
     public String getBootstrapServers() {
         return String.format("PLAINTEXT://%s:%s", getHost(), getMappedPort(KAFKA_PORT));
     }
 
+    public static String generateClusterId() {
+        UUID metadataTopicUuid = new UUID(0, 1L);
+        UUID ZeroUuid = new UUID(0, 0L);
+
+        UUID candidate = UUID.randomUUID();
+        while (candidate.equals(metadataTopicUuid) || candidate.equals(ZeroUuid)) {
+            candidate = UUID.randomUUID();
+        }
+
+        //From Java UUID to Kafka Uuid representation
+        ByteBuffer uuidBytes = ByteBuffer.wrap(new byte[16]);
+        uuidBytes.putLong(candidate.getMostSignificantBits());
+        uuidBytes.putLong(candidate.getLeastSignificantBits());
+        final byte[] uuidBytesArray = uuidBytes.array();
+
+        return Base64.getUrlEncoder().withoutPadding().encodeToString(uuidBytesArray);
+    }
```

Comment:
instead of duplicate logic from kafka client we can use `Uuid.randomUuid().toString()` at `KafkaContainerKraftCluster`

Reply:
That would imply to add a compile dependency.

Reply:
sorry, I don't mean using it in KafkaContainer but in the example. KafkaContainer just can receive the cluster id. This is related to the comment about make cluster id optional